src/core/base_bitset_type.hpp

```diff
@@ -0,0 +1,140 @@
+/*
+ * This file is part of OpenTTD.
+ * OpenTTD is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/**
+ * @file base_bitset_type.hpp Base for bitset types that accept strong types,
+ * i.e. types that need some casting like StrongType and enum class.
+ */
+
+#ifndef BASE_BITSET_TYPE_HPP
+#define BASE_BITSET_TYPE_HPP
+
+/**
+ * Base for bit set wrapper.
+ * Allows wrapping strong type values as a bit set. Methods are loosely modelled on std::bitset.
+ * @tparam Tvalue_type Type of values to wrap.
+ * @tparam Tstorage Storage type required to hold values.
+ */
+template <typename Timpl, typename Tvalue_type, typename Tstorage>
+class BaseBitSet {
+public:
+	using ValueType = Tvalue_type; ///< Value type of this BaseBitSet.
+	using BaseType = Tstorage; ///< Storage type of this BaseBitSet, be ConvertibleThroughBase
+
+	constexpr BaseBitSet() : data(0) {}
+	explicit constexpr BaseBitSet(Tstorage data) : data(data) {}
```

Comment:
Is `CompanyMask mask{CompanyID}` prevented here?

Reply:
No, but yes. Since everything needs to be a subclass of this base (CRTP), `CompanyMask` is its own class. That doesn't have a value constructor, so making anything but an empty `CompanyMask` is not possible.

Once `CompanyID` has become a `PoolID`, it could be considered adding a `ConvertibleThroughBaseBitset` type that would guarantee such a thing. And `CompanyMask`could then just use that class. That's something for later though.

Reply:
Cool, just checking that we can't accidentally assign the wrong thing here.